File: src/HaskellMobile/FilesDir.hs
{-# LANGUAGE ForeignFunctionInterface #-}
module HaskellMobile.FilesDir
  ( getAppFilesDir
  ) where

import Foreign.C.String (CString, peekCString)

foreign import ccall "getAppFilesDir" c_getAppFilesDir :: IO CString

-- | Returns the platform-specific app files directory path.
--
-- On Android this is the result of @getFilesDir().getAbsolutePath()@,
-- on iOS it is the Application Support directory, and on desktop it
-- falls back to @"."@ (the current working directory).
getAppFilesDir :: IO FilePath
getAppFilesDir = peekCString =<< c_getAppFilesDir

File: cbits/files_dir.c
/*
 * App files directory path.
 *
 * Uses a set/get pattern: platform bridges call setAppFilesDir() during
 * initialization, and Haskell code reads it via getAppFilesDir().
 *
 * Android: jni_bridge.c calls setAppFilesDir() in JNI_OnLoad
 *          (before haskellRunMain) via ActivityThread.currentApplication().
 * iOS:     UIBridgeIOS.m calls setAppFilesDir() in setup_ios_platform_globals
 *          (before haskellRunMain) via NSSearchPathForDirectoriesInDomains.
 * Desktop: falls back to "." (current working directory).
 */

#include <stdlib.h>

static const char *g_app_files_dir = NULL;

void setAppFilesDir(const char *path)
{
    g_app_files_dir = path;  /* caller owns the memory (static/strdup'd) */
}

const char* getAppFilesDir(void)
{
    if (g_app_files_dir) return g_app_files_dir;
    return ".";
}
